<html><head></head><body>
<div id="_idContainer039" class="calibre2">
<h1 class="chapter-number" id="_idParaDest-43"><a id="_idTextAnchor053" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.1.1">2</span></h1>
<h1 id="_idParaDest-44" class="calibre5"><a id="_idTextAnchor054" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.2.1">Understanding the Key Components of Genetic Algorithms</span></h1>
<p class="calibre3"><span class="kobospan" id="kobo.3.1">In this chapter, we will dive deeper into the key components and the implementation details of genetic algorithms in preparation for the following chapters, where we will use genetic algorithms to create solutions for various types </span><span><span class="kobospan" id="kobo.4.1">of problems.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.5.1">First, we will outline the basic flow of a genetic algorithm, and then break it down into its different components while demonstrating various implementations of selection methods, crossover methods, and mutation methods. </span><span class="kobospan" id="kobo.5.2">Next, we will look into real-coded genetic algorithms, which facilitate search in a continuous parameter space. </span><span class="kobospan" id="kobo.5.3">This will be followed by an overview of the intriguing topics of elitism, niching, and sharing in genetic algorithms. </span><span class="kobospan" id="kobo.5.4">Finally, we will study the art of solving problems using </span><span><span class="kobospan" id="kobo.6.1">genetic algorithms.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.7.1">By the end of this chapter, you will be able to do </span><span><span class="kobospan" id="kobo.8.1">the following:</span></span></p>
<ul class="calibre10">
<li class="calibre11"><span class="kobospan" id="kobo.9.1">Be familiar with the key components of </span><span><span class="kobospan" id="kobo.10.1">genetic algorithms</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.11.1">Understand the stages of the genetic </span><span><span class="kobospan" id="kobo.12.1">algorithm flow</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.13.1">Understand the genetic operators and become familiar with several of </span><span><span class="kobospan" id="kobo.14.1">their variants</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.15.1">Know the various options for </span><span><span class="kobospan" id="kobo.16.1">stopping conditions</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.17.1">Understand what modifications need to be made to the basic genetic algorithm when it’s applied to </span><span><span class="kobospan" id="kobo.18.1">real numbers</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.19.1">Understand the mechanism </span><span><span class="kobospan" id="kobo.20.1">of elitism</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.21.1">Understand the concepts and implementation of niching </span><span><span class="kobospan" id="kobo.22.1">and sharing</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.23.1">Know the choices you have to make when you’re starting to work on a </span><span><span class="kobospan" id="kobo.24.1">new problem</span></span></li>
</ul>
<h1 id="_idParaDest-45" class="calibre5"><a id="_idTextAnchor055" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.25.1">The basic flow of a genetic algorithm</span></h1>
<p class="calibre3"><span class="kobospan" id="kobo.26.1">The main</span><a id="_idIndexMarker038" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.27.1"> stages of the basic genetic algorithm flow are shown in the </span><span><span class="kobospan" id="kobo.28.1">following flowchart:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer015">
<span class="kobospan" id="kobo.29.1"><img alt="Figure 2.1: The basic flow of a genetic algorithm" src="image/B20851_02_001.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.30.1">Figure 2.1: The basic flow of a genetic algorithm</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.31.1">These stages are described in detail in the </span><span><span class="kobospan" id="kobo.32.1">following sections.</span></span></p>
<h2 id="_idParaDest-46" class="calibre7"><a id="_idTextAnchor056" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.33.1">Creating the initial population</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.34.1">The </span><a id="_idIndexMarker039" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.35.1">initial population is a set of valid candidate solutions (individuals) that are chosen randomly. </span><span class="kobospan" id="kobo.35.2">Since genetic algorithms use a chromosome to represent</span><a id="_idTextAnchor057" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.36.1"> each individual, the initial population is a set of chromosomes. </span><span class="kobospan" id="kobo.36.2">These chromosomes should conform to the chromosome format that we chose for the problem at hand – for example, binary strings of a </span><span><span class="kobospan" id="kobo.37.1">certain length.</span></span></p>
<h2 id="_idParaDest-47" class="calibre7"><a id="_idTextAnchor058" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.38.1">Calculating the fitness</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.39.1">The value of the </span><a id="_idIndexMarker040" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.40.1">fitness function is calculated fo</span><a id="_idTextAnchor059" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.41.1">r each individual. </span><span class="kobospan" id="kobo.41.2">This is done once for the initial population, and then for every new generation after applying the genetic operators of selection, crossover, and mutation. </span><span class="kobospan" id="kobo.41.3">As the fitness o</span><a id="_idTextAnchor060" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.42.1">f each individual is independent of the others, this calculation can be </span><span><span class="kobospan" id="kobo.43.1">done concurrently.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.44.1">Since the selection stage that follows the fitness calculation usually considers individuals with higher fitness scores to be better solutions, genetic algorithms are naturally geared toward finding the maximum value(s) of the fitness function. </span><span class="kobospan" id="kobo.44.2">If we have a problem where the minimum value is desired, the fitness calculation should inverse the original value – for example, by multiplying it by a value </span><span><span class="kobospan" id="kobo.45.1">of (-1).</span></span></p>
<h2 id="_idParaDest-48" class="calibre7"><a id="_idTextAnchor061" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.46.1">Applying selection, crossover, and mutation</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.47.1">Applying the genetic operators of selection, crossover, and mutation to the population results in the creation of a new generation that is based on better individuals than the </span><span><span class="kobospan" id="kobo.48.1">current ones.</span></span></p>
<ul class="calibre10">
<li class="calibre11"><span class="kobospan" id="kobo.49.1">The </span><strong class="bold"><span class="kobospan" id="kobo.50.1">selection</span></strong><span class="kobospan" id="kobo.51.1"> operator</span><a id="_idIndexMarker041" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.52.1"> is responsible for selecting </span><a id="_idIndexMarker042" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.53.1">individuals from the current population in a way that gives an advantage to better individuals. </span><span class="kobospan" id="kobo.53.2">Examples of selection operators are given in the </span><em class="italic"><span class="kobospan" id="kobo.54.1">Selection </span></em><span><em class="italic"><span class="kobospan" id="kobo.55.1">methods</span></em></span><span><span class="kobospan" id="kobo.56.1"> section.</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.57.1">The </span><strong class="bold"><span class="kobospan" id="kobo.58.1">crossover</span></strong><span class="kobospan" id="kobo.59.1"> (or </span><strong class="bold"><span class="kobospan" id="kobo.60.1">recombination</span></strong><span class="kobospan" id="kobo.61.1">) operator </span><a id="_idIndexMarker043" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.62.1">creates </span><a id="_idIndexMarker044" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.63.1">offspring from the selected individuals. </span><span class="kobospan" id="kobo.63.2">This is usually done by taking two selected individuals at a time and interchanging parts of their chromosomes to create two new chromosomes representing the offspring. </span><span class="kobospan" id="kobo.63.3">Examples of selection operators are given in the </span><em class="italic"><span class="kobospan" id="kobo.64.1">Crossover </span></em><span><em class="italic"><span class="kobospan" id="kobo.65.1">methods</span></em></span><span><span class="kobospan" id="kobo.66.1"> section.</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.67.1">The </span><strong class="bold"><span class="kobospan" id="kobo.68.1">mutation</span></strong><span class="kobospan" id="kobo.69.1"> operator</span><a id="_idIndexMarker045" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.70.1"> can randomly introduce a change</span><a id="_idIndexMarker046" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.71.1"> to one or more of the chromosome values (genes) of each newly created individual. </span><span class="kobospan" id="kobo.71.2">The mutation usually occurs with a very low probability. </span><span class="kobospan" id="kobo.71.3">Examples of mutation operators are given in the </span><em class="italic"><span class="kobospan" id="kobo.72.1">Mutation </span></em><span><em class="italic"><span class="kobospan" id="kobo.73.1">methods</span></em></span><span><span class="kobospan" id="kobo.74.1"> section.</span></span></li>
</ul>
<h2 id="_idParaDest-49" class="calibre7"><a id="_idTextAnchor062" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.75.1">Checking the stopping conditions</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.76.1">There can be multiple </span><a id="_idIndexMarker047" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.77.1">conditions to check against when determining whether the process can stop.</span><a id="_idTextAnchor063" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.78.1"> The two most commonly used stopping conditions are </span><span><span class="kobospan" id="kobo.79.1">as follows:</span></span></p>
<ul class="calibre10">
<li class="calibre11"><span class="kobospan" id="kobo.80.1">A maximum number of generations has been reached. </span><span class="kobospan" id="kobo.80.2">This also serves to limit the runtime and computing resources that are consumed by </span><span><span class="kobospan" id="kobo.81.1">the algorithm.</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.82.1">There was no noticeable improvement over the last few generations. </span><span class="kobospan" id="kobo.82.2">This can be implemented by storing the best fitness value that was achieved at every generation and comparing the current best value to the one that was achieved a predefined number of generations ago. </span><span class="kobospan" id="kobo.82.3">If the difference is smaller than a certain threshold, the algorithm </span><span><span class="kobospan" id="kobo.83.1">can stop.</span></span></li>
</ul>
<p class="calibre3"><span class="kobospan" id="kobo.84.1">Here are some other possible </span><span><span class="kobospan" id="kobo.85.1">stopping conditions:</span></span></p>
<ul class="calibre10">
<li class="calibre11"><span class="kobospan" id="kobo.86.1">The performance of the current best individual meets or exceeds the requirements of the specific </span><span><span class="kobospan" id="kobo.87.1">use case</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.88.1">A predetermined amount of time has elapsed since the </span><span><span class="kobospan" id="kobo.89.1">process began</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.90.1">A certain cost or budget has been consumed, such as CPU time </span><span><span class="kobospan" id="kobo.91.1">and/or memory</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.92.1">The best solution has taken over a portion of the population that is larger than a </span><span><span class="kobospan" id="kobo.93.1">preset threshold</span></span></li>
</ul>
<p class="calibre3"><span class="kobospan" id="kobo.94.1">To summarize, the genetic algorithm flow starts with a population of randomly generated candidate solutions (individuals), which are evaluated against the fitness function. </span><span class="kobospan" id="kobo.94.2">The heart of the flow is a loop where the genetic operators of selection, crossover, and mutation are applied successively, followed by re-evaluation of the individuals. </span><span class="kobospan" id="kobo.94.3">The loop continues until a stopping condition is encountered, upon which the best individual of the </span><a id="_idIndexMarker048" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.95.1">existing population is selected as our solution. </span><span class="kobospan" id="kobo.95.2">Now, let’s look at </span><span><span class="kobospan" id="kobo.96.1">selection methods.</span></span></p>
<h1 id="_idParaDest-50" class="calibre5"><a id="_idTextAnchor064" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.97.1">Selection methods</span></h1>
<p class="calibre3"><span class="kobospan" id="kobo.98.1">Selection</span><a id="_idIndexMarker049" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.99.1"> is used at the beginning of each cycle of the genetic algorithm flow to pick individuals from the current population that will be used as parents for the individuals of the next generation. </span><span class="kobospan" id="kobo.99.2">The selection is probability-based, and the probability of an individual being picked is tied to its fitness value, in a way that it gives an advantage to individuals with higher </span><span><span class="kobospan" id="kobo.100.1">fitness values.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.101.1">The following sections describe some of the </span><a id="_idIndexMarker050" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.102.1">commonly used selection methods and </span><span><span class="kobospan" id="kobo.103.1">their characteristics.</span></span></p>
<h2 id="_idParaDest-51" class="calibre7"><a id="_idTextAnchor065" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.104.1">Roulette wheel selection</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.105.1">In the roulette </span><a id="_idIndexMarker051" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.106.1">wheel selection method, also known as </span><strong class="bold"><span class="kobospan" id="kobo.107.1">fitness proportionate selection</span></strong><span class="kobospan" id="kobo.108.1"> (</span><strong class="bold"><span class="kobospan" id="kobo.109.1">FPS</span></strong><span class="kobospan" id="kobo.110.1">), the probability of selecting an </span><a id="_idIndexMarker052" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.111.1">individual is directly</span><a id="_idIndexMarker053" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.112.1"> proportionate to its fitness value. </span><span class="kobospan" id="kobo.112.2">This is comparable to using a roulette wheel in a casino and assigning each individual a portion of the wheel proportional to its fitness value. </span><span class="kobospan" id="kobo.112.3">When the wheel is turned, the odds of each individual being selected are proportional to the size of the portion of the wheel that </span><span><span class="kobospan" id="kobo.113.1">it occupies.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.114.1">For example, suppose we have a population of six individuals with fitness values, as shown in the following table. </span><span class="kobospan" id="kobo.114.2">The relative portion of the roulette wheel dedicated to each individual is calculated based on these </span><span><span class="kobospan" id="kobo.115.1">fitness values:</span></span></p>
<table class="no-table-style" id="table001-2">
<colgroup class="calibre12">
<col class="calibre13"/>
<col class="calibre13"/>
<col class="calibre13"/>
</colgroup>
<tbody class="calibre14">
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span><strong class="bold"><span class="kobospan" id="kobo.116.1">Individual</span></strong></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><strong class="bold" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.117.1">Fitness</span></strong></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><strong class="bold" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.118.1">Relative Portion</span></strong></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.119.1">A</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.120.1">8</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.121.1">7%</span></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.122.1">B</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.123.1">12</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.124.1">11%</span></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.125.1">C</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.126.1">27</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.127.1">24%</span></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.128.1">D</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.129.1">4</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.130.1">3%</span></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.131.1">E</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.132.1">45</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.133.1">40%</span></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.134.1">F</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.135.1">17</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.136.1">15%</span></span></p>
</td>
</tr>
</tbody>
</table>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.137.1">Table 2.1: Fitness value table</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.138.1">The </span><a id="_idIndexMarker054" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.139.1">matching roulette wheel is depicted in the </span><span><span class="kobospan" id="kobo.140.1">following diagram:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer016">
<span class="kobospan" id="kobo.141.1"><img alt="" role="presentation" src="image/B20851_02_002.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.142.1">Figure 2.2: Roulette wheel selection example</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.143.1">Each time the </span><a id="_idIndexMarker055" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.144.1">wheel is turned, the selection point is used to choose a single individual from the entire population. </span><span class="kobospan" id="kobo.144.2">The wheel is then turned again to select the next individual until we have enough individuals selected to fill the next generation. </span><span class="kobospan" id="kobo.144.3">As a result, the same individual can be picked </span><span><span class="kobospan" id="kobo.145.1">several times.</span></span></p>
<h2 id="_idParaDest-52" class="calibre7"><a id="_idTextAnchor066" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.146.1">Stochastic universal sampling</span></h2>
<p class="calibre3"><strong class="bold"><span class="kobospan" id="kobo.147.1">Stochastic universal sampling</span></strong><span class="kobospan" id="kobo.148.1"> (</span><strong class="bold"><span class="kobospan" id="kobo.149.1">SUS</span></strong><span class="kobospan" id="kobo.150.1">) is </span><a id="_idIndexMarker056" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.151.1">a slightly modified </span><a id="_idIndexMarker057" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.152.1">version of the roulette wheel selection described previously. </span><span class="kobospan" id="kobo.152.2">The same roulette wheel is used, with the same proportions, but instead of using a single selection point and turning the roulette wheel again and again until all needed individuals have been selected, we turn the wheel only once and use multiple selection points that are equally spaced around the wheel. </span><span class="kobospan" id="kobo.152.3">This way, all the individuals are chosen at the same time, as depicted in the </span><span><span class="kobospan" id="kobo.153.1">following diagram:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer017">
<span class="kobospan" id="kobo.154.1"><img alt="" role="presentation" src="image/B20851_02_003.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.155.1">Figure 2.3: SUS example</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.156.1">This selection method prevents individuals with particularly high fitness values from saturating the next generation by getting chosen over and over again too many times. </span><span class="kobospan" id="kobo.156.2">So, it provides weaker individuals with a chance to be chosen, reducing the somewhat unfair nature of the original roulette wheel </span><span><span class="kobospan" id="kobo.157.1">selection method.</span></span></p>
<h2 id="_idParaDest-53" class="calibre7"><a id="_idTextAnchor067" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.158.1">Rank-based selection</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.159.1">The rank-based selection</span><a id="_idIndexMarker058" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.160.1"> method is similar to roulette wheel </span><a id="_idIndexMarker059" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.161.1">selection, but instead of directly using the fitness values to calculate the probabilities for selecting each individual, the fitness is used just to sort the individuals. </span><span class="kobospan" id="kobo.161.2">Once sorted, each individual is given a rank representing their position, and the roulette probabilities are calculated based on </span><span><span class="kobospan" id="kobo.162.1">these ranks.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.163.1">For example, let’s take the same population of six individuals we previously used with the same fitness values. </span><span class="kobospan" id="kobo.163.2">To that, we’ll add the rank of each individual. </span><span class="kobospan" id="kobo.163.3">As the population size in our example is six, the highest-ranking individual gets a rank value of 6, the next one gets a rank value of 5, and so on. </span><span class="kobospan" id="kobo.163.4">The relative portion of the roulette wheel dedicated to each individual is now calculated based on these rank values instead of using the </span><span><span class="kobospan" id="kobo.164.1">fitness values:</span></span></p>
<table class="no-table-style" id="table002">
<colgroup class="calibre12">
<col class="calibre13"/>
<col class="calibre13"/>
<col class="calibre13"/>
<col class="calibre13"/>
</colgroup>
<tbody class="calibre14">
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span><strong class="bold"><span class="kobospan" id="kobo.165.1">Individual</span></strong></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><strong class="bold"><span class="kobospan" id="kobo.166.1">Fitness</span></strong></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><strong class="bold"><span class="kobospan" id="kobo.167.1">Rank</span></strong></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><strong class="bold"><span class="kobospan" id="kobo.168.1">Relative Portion</span></strong></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.169.1">A</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.170.1">8</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.171.1">2</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.172.1">9%</span></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.173.1">B</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.174.1">12</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.175.1">3</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.176.1">14%</span></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.177.1">C</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.178.1">27</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.179.1">5</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.180.1">24%</span></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.181.1">D</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.182.1">4</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.183.1">1</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.184.1">5%</span></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.185.1">E</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.186.1">45</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.187.1">6</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.188.1">29%</span></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.189.1">F</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.190.1">17</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.191.1">4</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.192.1">19%</span></span></p>
</td>
</tr>
</tbody>
</table>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.193.1">Table 2.2: Fitness value table based on the relative portion</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.194.1">The matching roulette wheel is depicted in the </span><span><span class="kobospan" id="kobo.195.1">following diagram:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer018">
<span class="kobospan" id="kobo.196.1"><img alt="" role="presentation" src="image/B20851_02_004.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.197.1">Figure 2.4: Rank-based selection example</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.198.1">Rank-based selection can be useful when a few individuals have much larger fitness values than all the rest. </span><span class="kobospan" id="kobo.198.2">Using rank instead of raw fitness prevents these few individuals from taking over the entire population of the next generation as ranking eliminates the </span><span><span class="kobospan" id="kobo.199.1">large differences.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.200.1">Another </span><a id="_idIndexMarker060" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.201.1">useful case is when all individuals have</span><a id="_idIndexMarker061" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.202.1"> similar fitness values. </span><span class="kobospan" id="kobo.202.2">Here, rank-based selection will spread them apart, giving a clearer advantage to the better ones, even if the fitness differences </span><span><span class="kobospan" id="kobo.203.1">are small.</span></span></p>
<h2 id="_idParaDest-54" class="calibre7"><a id="_idTextAnchor068" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.204.1">Fitness scaling</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.205.1">While rank-based selection</span><a id="_idIndexMarker062" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.206.1"> replaces each fitness value with the </span><a id="_idIndexMarker063" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.207.1">individual’s rank, fitness scaling applies a scaling transformation to the raw fitness values and replaces them with the transformation’s result. </span><span class="kobospan" id="kobo.207.2">The transformation maps the raw fitness values into a desired range, </span><span><span class="kobospan" id="kobo.208.1">as follows:</span></span></p>
<p class="calibre3"><em class="italic"><span class="kobospan" id="kobo.209.1">scaled fitness = a × (raw fitness) + b</span></em></p>
<p class="calibre3"><span class="kobospan" id="kobo.210.1">Here, </span><em class="italic"><span class="kobospan" id="kobo.211.1">a</span></em><span class="kobospan" id="kobo.212.1"> and </span><em class="italic"><span class="kobospan" id="kobo.213.1">b</span></em><span class="kobospan" id="kobo.214.1"> are constants that we can select to achieve the desired range of the </span><span><span class="kobospan" id="kobo.215.1">scaled fitness.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.216.1">For example, if we use the same values from the previous examples, the range of the raw fitness values is between 4 (lowest fitness value, individual D) and 45 (highest fitness value, individual E). </span><span class="kobospan" id="kobo.216.2">Suppose we want to map the values into a new range, between 50 and 100. </span><span class="kobospan" id="kobo.216.3">We can calculate the values of the </span><em class="italic"><span class="kobospan" id="kobo.217.1">a</span></em><span class="kobospan" id="kobo.218.1"> and </span><em class="italic"><span class="kobospan" id="kobo.219.1">b</span></em><span class="kobospan" id="kobo.220.1"> constants using the following equations, representing these </span><span><span class="kobospan" id="kobo.221.1">two individuals:</span></span></p>
<ul class="calibre10">
<li class="calibre11"><em class="italic"><span class="kobospan" id="kobo.222.1">50 = a × 4 + b</span></em><span class="kobospan" id="kobo.223.1"> (lowest </span><span><span class="kobospan" id="kobo.224.1">fitness value)</span></span></li>
<li class="calibre11"><em class="italic"><span class="kobospan" id="kobo.225.1">100 = a × 45 + b</span></em><span class="kobospan" id="kobo.226.1"> (highest </span><span><span class="kobospan" id="kobo.227.1">fitness value)</span></span></li>
</ul>
<p class="calibre3"><span class="kobospan" id="kobo.228.1">Solving this simple system of linear equations will yield the following scaling </span><span><span class="kobospan" id="kobo.229.1">parameter values:</span></span></p>
<p class="calibre3"><em class="italic"><span class="kobospan" id="kobo.230.1">a = 1.22, b = </span></em><span><em class="italic"><span class="kobospan" id="kobo.231.1">45.12</span></em></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.232.1">This means that the scaled fitness values can be calculated </span><span><span class="kobospan" id="kobo.233.1">as follows:</span></span></p>
<p class="calibre3"><em class="italic"><span class="kobospan" id="kobo.234.1">scaled fitness = 1.22 × (raw fitness) + </span></em><span><em class="italic"><span class="kobospan" id="kobo.235.1">45.12</span></em></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.236.1">After </span><a id="_idIndexMarker064" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.237.1">adding a new column to the table containing</span><a id="_idIndexMarker065" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.238.1"> the scaled fitness values, we can see that the range is indeed between 50 and 100, </span><span><span class="kobospan" id="kobo.239.1">as desired:</span></span></p>
<table class="no-table-style" id="table003">
<colgroup class="calibre12">
<col class="calibre13"/>
<col class="calibre13"/>
<col class="calibre13"/>
<col class="calibre13"/>
</colgroup>
<tbody class="calibre14">
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span><strong class="bold"><span class="kobospan" id="kobo.240.1">Individual</span></strong></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><strong class="bold"><span class="kobospan" id="kobo.241.1">Fitness</span></strong></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><strong class="bold"><span class="kobospan" id="kobo.242.1">Scaled Fitness</span></strong></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><strong class="bold"><span class="kobospan" id="kobo.243.1">Relative Portion</span></strong></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.244.1">A</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.245.1">8</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.246.1">55</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.247.1">13%</span></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.248.1">B</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.249.1">12</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.250.1">60</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.251.1">15%</span></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.252.1">C</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.253.1">27</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.254.1">78</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.255.1">19%</span></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.256.1">D</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.257.1">4</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.258.1">50</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.259.1">12%</span></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.260.1">E</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.261.1">45</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.262.1">100</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.263.1">25%</span></span></p>
</td>
</tr>
<tr class="no-table-style1">
<td class="no-table-style2">
<p class="calibre3"><span class="kobospan" id="kobo.264.1">F</span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.265.1">17</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.266.1">66</span></span></p>
</td>
<td class="no-table-style2">
<p class="calibre3"><span><span class="kobospan" id="kobo.267.1">16%</span></span></p>
</td>
</tr>
</tbody>
</table>
<p class="calibre3"><span class="kobospan" id="kobo.268.1">The matching roulette wheel is depicted in the </span><span><span class="kobospan" id="kobo.269.1">following diagram:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer019">
<span class="kobospan" id="kobo.270.1"><img alt="" role="presentation" src="image/B20851_02_005.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.271.1">Figure 2.5: Roulette wheel selection example after fitness scaling</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.272.1">As </span><span><em class="italic"><span class="kobospan" id="kobo.273.1">Figure 2</span></em></span><em class="italic"><span class="kobospan" id="kobo.274.1">.5</span></em><span class="kobospan" id="kobo.275.1"> illustrates, scaling the</span><a id="_idIndexMarker066" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.276.1"> fitness values to the new range provided a</span><a id="_idIndexMarker067" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.277.1"> much more moderate partition of the roulette wheel compared to the original partition. </span><span class="kobospan" id="kobo.277.2">The best individual (with a scaled fitness value of 100) is now only twice more likely to be selected than the worst one (with a scaled fitness value of 50), instead of being more than 11 times more likely to be chosen when using the raw </span><span><span class="kobospan" id="kobo.278.1">fitness values.</span></span></p>
<h2 id="_idParaDest-55" class="calibre7"><a id="_idTextAnchor069" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.279.1">Tournament selection</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.280.1">In each round of the </span><a id="_idIndexMarker068" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.281.1">tournament selection method, two or</span><a id="_idIndexMarker069" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.282.1"> more individuals are randomly picked from the population, and the one with the highest fitness score wins and </span><span><span class="kobospan" id="kobo.283.1">gets selected.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.284.1">For example, suppose we have the same six individuals and the same fitness values we used in the previous examples. </span><span class="kobospan" id="kobo.284.2">The following diagram illustrates randomly selecting three of them (A, B, and F), then announcing F as the winner since it has the largest fitness value (17) among these </span><span><span class="kobospan" id="kobo.285.1">three individuals:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer020">
<span class="kobospan" id="kobo.286.1"><img alt="Figure 2.6: Tournament selection example with a tournament size of three" src="image/B20851_02_006.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.287.1">Figure 2.6: Tournament selection example with a tournament size of three</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.288.1">The </span><a id="_idIndexMarker070" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.289.1">number of individuals participating at each tournament </span><a id="_idIndexMarker071" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.290.1">selection round (three in our example) is suitably</span><a id="_idIndexMarker072" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.291.1"> called </span><strong class="bold"><span class="kobospan" id="kobo.292.1">tournament size</span></strong><span class="kobospan" id="kobo.293.1">. </span><span class="kobospan" id="kobo.293.2">The larger the tournament size, the higher the chances that the best individuals will participate in the tournaments, and the lesser the chances of low-scoring individuals winning a tournament and </span><span><span class="kobospan" id="kobo.294.1">getting selected.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.295.1">An interesting aspect of this selection method is that, so long as we can compare any two individuals and determine which of them is better, the actual value of the fitness function is not needed. </span><span class="kobospan" id="kobo.295.2">Next, we’ll look at </span><span><span class="kobospan" id="kobo.296.1">crossover methods.</span></span></p>
<h1 id="_idParaDest-56" class="calibre5"><a id="_idTextAnchor070" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.297.1">Crossover methods</span></h1>
<p class="calibre3"><span class="kobospan" id="kobo.298.1">The</span><a id="_idIndexMarker073" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.299.1"> crossover operator, also referred to as recombination, corresponds to the crossover that takes place during sexual reproduction in biology and is used to combine the genetic information of two individuals, serving as parents, to produce (usually </span><span><span class="kobospan" id="kobo.300.1">two) offspring.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.301.1">The crossover operator is typically applied with some (high) probability value. </span><span class="kobospan" id="kobo.301.2">Whenever crossover is </span><em class="italic"><span class="kobospan" id="kobo.302.1">not</span></em><span class="kobospan" id="kobo.303.1"> applied, both parents are directly cloned into the </span><span><span class="kobospan" id="kobo.304.1">next generation.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.305.1">The following sections describe some of the commonly used crossover methods and their typical use cases. </span><span class="kobospan" id="kobo.305.2">However, in certain situations, you may opt to use a problem-specific crossover method</span><a id="_idIndexMarker074" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.306.1"> that will be more suitable for a </span><span><span class="kobospan" id="kobo.307.1">particular case.</span></span></p>
<h2 id="_idParaDest-57" class="calibre7"><a id="_idTextAnchor071" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.308.1">Single-point crossover</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.309.1">In the single-point</span><a id="_idIndexMarker075" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.310.1"> crossover method, a location on the chromosomes of</span><a id="_idIndexMarker076" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.311.1"> both parents is selected randomly. </span><span class="kobospan" id="kobo.311.2">This location is referred to</span><a id="_idIndexMarker077" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.312.1"> as the </span><strong class="bold"><span class="kobospan" id="kobo.313.1">crossover point</span></strong><span class="kobospan" id="kobo.314.1"> or </span><strong class="bold"><span class="kobospan" id="kobo.315.1">cut point</span></strong><span class="kobospan" id="kobo.316.1">. </span><span class="kobospan" id="kobo.316.2">Genes to</span><a id="_idIndexMarker078" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.317.1"> the right of that point are swapped between the two parent chromosomes. </span><span class="kobospan" id="kobo.317.2">As a result, we get two offspring, and each of them carries some genetic information from </span><span><span class="kobospan" id="kobo.318.1">both parents.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.319.1">The following diagram demonstrates a single-point crossover operation being conducted on a pair of binary chromosomes, with the crossover point located between the fifth and </span><span><span class="kobospan" id="kobo.320.1">sixth genes:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer021">
<span class="kobospan" id="kobo.321.1"><img alt="Figure 2.7: Single-point crossover example" src="image/B20851_02_007.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.322.1">Figure 2.7: Single-point crossover example</span></p>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.323.1">Source: </span><a href="https://commons.wikimedia.org/wiki/File:Computational.science.Genetic.algorithm.Crossover.One.Point.svg" class="calibre6 pcalibre pcalibre1"><span class="kobospan" id="kobo.324.1">https://commons.wikimedia.org/wiki/File:Computational.science.Genetic.algorithm.Crossover.One.Point.svg</span></a><span class="kobospan" id="kobo.325.1">. </span></p>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.326.1">Image by Yearofthedragon. </span></p>
<p class="calibre3"><span class="kobospan" id="kobo.327.1">In the next section, we will cover extensions of this method, namely two-point and </span><span><span class="kobospan" id="kobo.328.1">k-point crossover.</span></span></p>
<h2 id="_idParaDest-58" class="calibre7"><a id="_idTextAnchor072" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.329.1">Two-point and k-point crossover</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.330.1">In the two-point crossover</span><a id="_idIndexMarker079" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.331.1"> method, two </span><a id="_idIndexMarker080" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.332.1">crossover points on the chromosomes of both </span><a id="_idIndexMarker081" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.333.1">parents are</span><a id="_idIndexMarker082" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.334.1"> selected randomly. </span><span class="kobospan" id="kobo.334.2">The genes residing between these points are swapped between the two </span><span><span class="kobospan" id="kobo.335.1">parent chromosomes.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.336.1">The following diagram demonstrates a two-point crossover carried out on a pair of binary chromosomes, with the first crossover point located between the third and fourth genes, and the other between the seventh and </span><span><span class="kobospan" id="kobo.337.1">eighth genes:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer022">
<span class="kobospan" id="kobo.338.1"><img alt="Figure 2.8: Two-point crossover example" src="image/B20851_02_008.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.339.1">Figure 2.8: Two-point crossover example</span></p>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.340.1">Source: </span><a href="https://commons.wikimedia.org/wiki/File:Computational.science.Genetic.algorithm.Crossover.Two.Point.svg" class="calibre6 pcalibre pcalibre1"><span class="kobospan" id="kobo.341.1">https://commons.wikimedia.org/wiki/File:Computational.science.Genetic.algorithm.Crossover.Two.Point.svg</span></a><span class="kobospan" id="kobo.342.1">. </span></p>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.343.1">Image by Yearofthedragon.</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.344.1">The two-point crossover method can be implemented by carrying out two single-point crossovers, each with a different crossover point. </span><span class="kobospan" id="kobo.344.2">A generalization of this method is k-point crossover, where </span><em class="italic"><span class="kobospan" id="kobo.345.1">k</span></em><span class="kobospan" id="kobo.346.1"> represents a positive integer, and </span><em class="italic"><span class="kobospan" id="kobo.347.1">k</span></em><span class="kobospan" id="kobo.348.1"> crossover points </span><span><span class="kobospan" id="kobo.349.1">are used.</span></span></p>
<h2 id="_idParaDest-59" class="calibre7"><a id="_idTextAnchor073" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.350.1">Uniform crossover</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.351.1">In the uniform crossover</span><a id="_idIndexMarker083" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.352.1"> method, each gene is independently </span><a id="_idIndexMarker084" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.353.1">determined by randomly choosing one of the parents. </span><span class="kobospan" id="kobo.353.2">When the random distribution is 50%, each parent has the same chance of influencing the offspring, as illustrated in the </span><span><span class="kobospan" id="kobo.354.1">following diagram:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer023">
<span class="kobospan" id="kobo.355.1"><img alt="Figure 2.9: Uniform crossover example" src="image/B20851_02_009.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.356.1">Figure 2.9: Uniform crossover example</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.357.1">Note that, in </span><a id="_idIndexMarker085" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.358.1">this example, the second offspring was created by </span><a id="_idIndexMarker086" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.359.1">complementing the choices made for the first offspring. </span><span class="kobospan" id="kobo.359.2">However, both offspring can also be created independently of </span><span><span class="kobospan" id="kobo.360.1">each other.</span></span></p>
<p class="callout-heading"><span class="kobospan" id="kobo.361.1">Important note</span></p>
<p class="callout"><span class="kobospan" id="kobo.362.1">In this example, we used integer-based chromosomes, but it would work similarly with </span><span><span class="kobospan" id="kobo.363.1">binary ones.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.364.1">Since this method does not exchange entire segments of the chromosome, it has greater potential for diversity in the </span><span><span class="kobospan" id="kobo.365.1">resulting offspring.</span></span></p>
<h2 id="_idParaDest-60" class="calibre7"><a id="_idTextAnchor074" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.366.1">Crossover for ordered lists</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.367.1">In the previous </span><a id="_idIndexMarker087" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.368.1">example, we saw the results of a crossover operation on two integer-based chromosomes. </span><span class="kobospan" id="kobo.368.2">While each of the parents had every value between 0 and 9 appear exactly once, each of the resulting offspring had certain values appearing more than once (for example, 2 in the top offspring and 1 in the other), and other values were missing (such as 4 in the top offspring and 5 in </span><span><span class="kobospan" id="kobo.369.1">the other).</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.370.1">In some tasks, however, integer-based chromosomes may represent indices of an ordered list. </span><span class="kobospan" id="kobo.370.2">For example, suppose we have several cities; we know the distance between each, and we need to find the shortest possible route through all of them. </span><span class="kobospan" id="kobo.370.3">This is known as the traveling salesman problem and will be covered in detail in one of the </span><span><span class="kobospan" id="kobo.371.1">following chapters.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.372.1">If, for instance, we have</span><a id="_idIndexMarker088" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.373.1"> four cities, a convenient way to represent a possible solution for this problem would be a four-integer chromosome showing the order of visiting the cities – for example, (1,2,3,4) or (3,4,2,1). </span><span class="kobospan" id="kobo.373.2">A chromosome with two of the same values or missing one of the values, such as (1,2,2,4), will not represent a </span><span><span class="kobospan" id="kobo.374.1">valid solution.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.375.1">For such cases, alternative crossover methods were devised to ensure that the offspring that were created would still be valid. </span><span class="kobospan" id="kobo.375.2">One of these </span><a id="_idIndexMarker089" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.376.1">methods, </span><strong class="bold"><span class="kobospan" id="kobo.377.1">ordered crossover</span></strong><span class="kobospan" id="kobo.378.1"> (</span><strong class="bold"><span class="kobospan" id="kobo.379.1">OX1</span></strong><span class="kobospan" id="kobo.380.1">) will be covered in the </span><span><span class="kobospan" id="kobo.381.1">following section.</span></span></p>
<h3 class="calibre9"><span class="kobospan" id="kobo.382.1">OX1</span></h3>
<p class="calibre3"><span class="kobospan" id="kobo.383.1">The OX1 method</span><a id="_idIndexMarker090" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.384.1"> strives to preserve the relative ordering of the parent’s genes as much </span><a id="_idIndexMarker091" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.385.1">as possible. </span><span class="kobospan" id="kobo.385.2">We will demonstrate it by using chromosomes with a length </span><span><span class="kobospan" id="kobo.386.1">of six.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.387.1">In this example, we used integer-based chromosomes, but it would work similarly with </span><span><span class="kobospan" id="kobo.388.1">binary ones.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.389.1">The first step is a two-point crossover with random cut points, as shown in the following diagram (with the parents depicted on the </span><span><span class="kobospan" id="kobo.390.1">left-hand side):</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer024">
<span class="kobospan" id="kobo.391.1"><img alt="Figure 2.10: OX1 example – step 1" src="image/B20851_02_010.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.392.1">Figure 2.10: OX1 example – step 1</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.393.1">We will now start filling in the rest of the genes in each offspring by going over all the parent’s genes in their original order, starting after the second cut point. </span><span class="kobospan" id="kobo.393.2">For the first parent, we find a </span><em class="italic"><span class="kobospan" id="kobo.394.1">6</span></em><span class="kobospan" id="kobo.395.1">, but this is already present in the offspring, so we continue (with wrapping around) to </span><em class="italic"><span class="kobospan" id="kobo.396.1">1</span></em><span class="kobospan" id="kobo.397.1">; this is already present too. </span><span class="kobospan" id="kobo.397.2">The next in order is the </span><em class="italic"><span class="kobospan" id="kobo.398.1">2</span></em><span class="kobospan" id="kobo.399.1">. </span><span class="kobospan" id="kobo.399.2">Since </span><em class="italic"><span class="kobospan" id="kobo.400.1">2</span></em><span class="kobospan" id="kobo.401.1"> is not yet present in the offspring, we add it there, as shown in the following figure. </span><span class="kobospan" id="kobo.401.2">For the second parent-offspring pair, we start with the parent’s </span><em class="italic"><span class="kobospan" id="kobo.402.1">5</span></em><span class="kobospan" id="kobo.403.1">, which is already present in the offspring, then move on to </span><em class="italic"><span class="kobospan" id="kobo.404.1">4</span></em><span class="kobospan" id="kobo.405.1">, which is present as well, and end up with </span><em class="italic"><span class="kobospan" id="kobo.406.1">2</span></em><span class="kobospan" id="kobo.407.1">, which is not present yet and therefore gets added. </span><span class="kobospan" id="kobo.407.2">This is shown in the following diagram </span><span><span class="kobospan" id="kobo.408.1">as well:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer025">
<span class="kobospan" id="kobo.409.1"><img alt="Figure 2.11: OX1 example – step 2" src="image/B20851_02_011.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.410.1">Figure 2.11: OX1 example – step 2</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.411.1">For the top parent, we now continue to </span><em class="italic"><span class="kobospan" id="kobo.412.1">3</span></em><span class="kobospan" id="kobo.413.1"> (already present in the offspring), and then </span><em class="italic"><span class="kobospan" id="kobo.414.1">4</span></em><span class="kobospan" id="kobo.415.1">, which gets added to the offspring. </span><span class="kobospan" id="kobo.415.2">For the other parent, the next gene is </span><em class="italic"><span class="kobospan" id="kobo.416.1">6</span></em><span class="kobospan" id="kobo.417.1">. </span><span class="kobospan" id="kobo.417.2">Since it’s not present in the matching offspring, it gets added to it. </span><span class="kobospan" id="kobo.417.3">The results are illustrated in the </span><span><span class="kobospan" id="kobo.418.1">following diagram:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer026">
<span class="kobospan" id="kobo.419.1"><img alt="Figure 2.12: OX1 example – step 3" src="image/B20851_02_012.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.420.1">Figure 2.12: OX1 example – step 3</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.421.1">We continue </span><a id="_idIndexMarker092" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.422.1">similarly with the next genes not yet present in the offspring and fill in the </span><a id="_idIndexMarker093" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.423.1">last available spots, as depicted in the </span><span><span class="kobospan" id="kobo.424.1">following diagram:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer027">
<span class="kobospan" id="kobo.425.1"><img alt="Figure 2.13: OX1 example – step 4" src="image/B20851_02_013.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.426.1">Figure 2.13: OX1 example – step 4</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.427.1">This completes the process of producing two valid offspring chromosomes, as the following </span><span><span class="kobospan" id="kobo.428.1">diagram demonstrates:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer028">
<span class="kobospan" id="kobo.429.1"><img alt="Figure 2.14: OX1 example – step 5" src="image/B20851_02_014.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.430.1">Figure 2.14: OX1 example – step 5</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.431.1">There are </span><a id="_idIndexMarker094" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.432.1">numerous other methods to implement crossover, some of which we</span><a id="_idIndexMarker095" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.433.1"> will encounter later in this book. </span><span class="kobospan" id="kobo.433.2">However, thanks to the versatility of genetic algorithms, you can always come up with your own methods. </span><span class="kobospan" id="kobo.433.3">In the next section, we’ll consider </span><span><span class="kobospan" id="kobo.434.1">mutation methods.</span></span></p>
<h1 id="_idParaDest-61" class="calibre5"><a id="_idTextAnchor075" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.435.1">Mutation methods</span></h1>
<p class="calibre3"><span class="kobospan" id="kobo.436.1">Mutation</span><a id="_idIndexMarker096" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.437.1"> is the last genetic operator to be applied in the process of creating a new generation. </span><span class="kobospan" id="kobo.437.2">The mutation operator is applied to the offspring that were created as a result of the selection and </span><span><span class="kobospan" id="kobo.438.1">crossover operations.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.439.1">The mutation operator is probability-based and usually occurs at a (very) low probability as it carries the risk of harming the performance of any individual it is applied to. </span><span class="kobospan" id="kobo.439.2">In some versions of genetic algorithms, the mutation probability gradually increases as the generations advance to prevent stagnation and ensure the diversity of the population. </span><span class="kobospan" id="kobo.439.3">On the other hand, if the mutation rate is increased excessively, the genetic algorithm will turn into the equivalent of a </span><span><span class="kobospan" id="kobo.440.1">random search.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.441.1">The following sections describe some of the commonly used mutation methods and their typical use cases. </span><span class="kobospan" id="kobo.441.2">However, remember that you can always choose to use your own problem-specific mutation method that you deem more suitable for your particular </span><span><span class="kobospan" id="kobo.442.1">use case.</span></span></p>
<h2 id="_idParaDest-62" class="calibre7"><a id="_idTextAnchor076" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.443.1">Flip-bit mutation</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.444.1">When applying </span><a id="_idIndexMarker097" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.445.1">flip-bit mutation</span><a id="_idIndexMarker098" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.446.1"> to a binary chromosome, one gene is randomly selected and its value is flipped (complemented), as shown in the </span><span><span class="kobospan" id="kobo.447.1">following diagram:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer029">
<span class="kobospan" id="kobo.448.1"><img alt="Figure 2.15: Flip-bit mutation example" src="image/B20851_02_015.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.449.1">Figure 2.15: Flip-bit mutation example</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.450.1">This</span><a id="_idIndexMarker099" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.451.1"> can be extended to several random genes being flipped instead of </span><a id="_idIndexMarker100" class="calibre6 pcalibre pcalibre1"/><span><span class="kobospan" id="kobo.452.1">just one.</span></span></p>
<h2 id="_idParaDest-63" class="calibre7"><a id="_idTextAnchor077" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.453.1">Swap mutation</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.454.1">When applying the </span><a id="_idIndexMarker101" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.455.1">swap mutation</span><a id="_idIndexMarker102" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.456.1"> to binary or integer-based chromosomes, two genes are randomly selected and their values are swapped, as shown in the </span><span><span class="kobospan" id="kobo.457.1">following diagram:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer030">
<span class="kobospan" id="kobo.458.1"><img alt="Figure 2.16: Swap mutation example" src="image/B20851_02_016.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.459.1">Figure 2.16: Swap mutation example</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.460.1">This mutation operation is suitable for the chromosomes of ordered lists as the new chromosome still carries the same genes as the </span><span><span class="kobospan" id="kobo.461.1">original one.</span></span></p>
<h2 id="_idParaDest-64" class="calibre7"><a id="_idTextAnchor078" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.462.1">Inversion mutation</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.463.1">When applying the </span><a id="_idIndexMarker103" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.464.1">inversion mutation</span><a id="_idIndexMarker104" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.465.1"> to binary or integer-based chromosomes, a random sequence of genes is selected and the order of the genes in that sequence is reversed, as depicted in the </span><span><span class="kobospan" id="kobo.466.1">following diagram:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer031">
<span class="kobospan" id="kobo.467.1"><img alt="Figure 2.17: Inversion mutation example" src="image/B20851_02_017.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.468.1">Figure 2.17: Inversion mutation example</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.469.1">Similar to the swap mutation, the inversion mutation operation is suitable for the chromosomes of </span><span><span class="kobospan" id="kobo.470.1">ordered lists.</span></span></p>
<h2 id="_idParaDest-65" class="calibre7"><a id="_idTextAnchor079" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.471.1">Scramble mutation</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.472.1">Another mutation operator </span><a id="_idIndexMarker105" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.473.1">that’s </span><a id="_idIndexMarker106" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.474.1">suitable for the chromosomes of ordered lists is the scramble mutation. </span><span class="kobospan" id="kobo.474.2">When applied, a random sequence of genes is selected and the order of the genes in that sequence is shuffled (or scrambled), </span><span><span class="kobospan" id="kobo.475.1">as follows:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer032">
<span class="kobospan" id="kobo.476.1"><img alt="Figure 2.18: Scramble mutation example" src="image/B20851_02_018.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.477.1">Figure 2.18: Scramble mutation example</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.478.1">In the next section, we will cover some other types of specialized operators that have been created for real-coded </span><span><span class="kobospan" id="kobo.479.1">genetic algorithms.</span></span></p>
<h1 id="_idParaDest-66" class="calibre5"><a id="_idTextAnchor080" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.480.1">Real-coded genetic algorithms</span></h1>
<p class="calibre3"><span class="kobospan" id="kobo.481.1">So far, we have seen</span><a id="_idIndexMarker107" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.482.1"> chromosomes that represent binary or integer parameters. </span><span class="kobospan" id="kobo.482.2">Consequently, the genetic operators were suitable for working on these types of chromosomes. </span><span class="kobospan" id="kobo.482.3">However, we often encounter problems where the solution space is continuous. </span><span class="kobospan" id="kobo.482.4">In other words, the individuals are made up of real (</span><span><span class="kobospan" id="kobo.483.1">floating-point) numbers.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.484.1">Historically, genetic algorithms used binary strings to represent integers as well as real numbers; however, this was not ideal. </span><span class="kobospan" id="kobo.484.2">The precision of a real number represented using a binary string is limited by the length of the string (number of bits). </span><span class="kobospan" id="kobo.484.3">Since we need to determine this length in advance, we may end up with binary strings that are too short, resulting in insufficient precision, or are </span><span><span class="kobospan" id="kobo.485.1">overly long.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.486.1">Moreover, when a binary string is used to represent a number, the significance of each bit varies by its location – the most significant bit being on the left. </span><span class="kobospan" id="kobo.486.2">This can cause imbalance related to schemas – the patterns occurring in the chromosomes. </span><span class="kobospan" id="kobo.486.3">For example, the schema 1**** (representing all five-digit binary strings starting with 1) and the schema ****1 (representing all five-digit binary strings ending with 1) both have an order of 1 and a defining length of 0; however, the first one carries much more significance than </span><span><span class="kobospan" id="kobo.487.1">the other.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.488.1">Instead of using binary strings, arrays of real-valued numbers were found to be a simpler and better approach. </span><span class="kobospan" id="kobo.488.2">For example, if we have a problem involving three real-valued parameters, the chromosome will look like </span><em class="italic"><span class="kobospan" id="kobo.489.1">[x1, x2, x3]</span></em><span class="kobospan" id="kobo.490.1">, where </span><em class="italic"><span class="kobospan" id="kobo.491.1">x1</span></em><span class="kobospan" id="kobo.492.1">, </span><em class="italic"><span class="kobospan" id="kobo.493.1">x2</span></em><span class="kobospan" id="kobo.494.1">, and </span><em class="italic"><span class="kobospan" id="kobo.495.1">x3</span></em><span class="kobospan" id="kobo.496.1"> represent real numbers, such as [1.23, 7.2134, -25.309] or [-30.10, </span><span><span class="kobospan" id="kobo.497.1">100.2, 42.424].</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.498.1">The various selection methods mentioned earlier in this chapter will work just the same for real-coded chromosomes as they only depend on the fitness of the individuals and not </span><span><span class="kobospan" id="kobo.499.1">their representation.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.500.1">However, the</span><a id="_idIndexMarker108" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.501.1"> crossover and mutation methods we’ve covered so far will not be suitable for the real-coded chromosomes, so specialized ones need to be used. </span><span class="kobospan" id="kobo.501.2">One important point to remember is that these crossover and mutation operations are applied separately for each dimension of the array that forms the real-coded chromosome. </span><span class="kobospan" id="kobo.501.3">For example, if [1.23, 7.213, -25.39] and [-30.10, 100.2, 42.42] are parents that have been selected for the crossover operation, the crossover will be done separately for the </span><span><span class="kobospan" id="kobo.502.1">following pairs:</span></span></p>
<ul class="calibre10">
<li class="calibre11"><span class="kobospan" id="kobo.503.1">1.23 and -30.10 (</span><span><span class="kobospan" id="kobo.504.1">first dimension)</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.505.1">7.213 and 100.2 (</span><span><span class="kobospan" id="kobo.506.1">second dimension)</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.507.1">-25.39 and 42.42 (</span><span><span class="kobospan" id="kobo.508.1">third dimension)</span></span></li>
</ul>
<p class="calibre3"><span class="kobospan" id="kobo.509.1">This is illustrated in the </span><span><span class="kobospan" id="kobo.510.1">following diagram:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer033">
<span class="kobospan" id="kobo.511.1"><img alt="Figure 2.19: Real-coded chromosomes crossover example" src="image/B20851_02_019.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.512.1">Figure 2.19: Real-coded chromosomes crossover example</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.513.1">Similarly, the mutation operator, when applied to a real-coded chromosome, will apply separately to </span><span><span class="kobospan" id="kobo.514.1">each dimension.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.515.1">Several of these real-coded crossover and mutation methods will be described in the following sections. </span><span class="kobospan" id="kobo.515.2">Later, in </span><a href="B20851_06.xhtml#_idTextAnchor197" class="calibre6 pcalibre pcalibre1"><span><em class="italic"><span class="kobospan" id="kobo.516.1">Chapter 6</span></em></span></a><span class="kobospan" id="kobo.517.1">, </span><em class="italic"><span class="kobospan" id="kobo.518.1">Optimizing Continuous Functions</span></em><span class="kobospan" id="kobo.519.1">, we will get to see them </span><span><span class="kobospan" id="kobo.520.1">in action.</span></span></p>
<h2 id="_idParaDest-67" class="calibre7"><a id="_idTextAnchor081" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.521.1">Blend crossover</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.522.1">In the </span><strong class="bold"><span class="kobospan" id="kobo.523.1">blend crossover</span></strong><span class="kobospan" id="kobo.524.1"> (</span><strong class="bold"><span class="kobospan" id="kobo.525.1">BLX</span></strong><span class="kobospan" id="kobo.526.1">) method, each </span><a id="_idIndexMarker109" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.527.1">offspring is randomly selected from the following interval created by </span><span><span class="kobospan" id="kobo.528.1">its </span></span><span><a id="_idIndexMarker110" class="calibre6 pcalibre pcalibre1"/></span><span><span class="kobospan" id="kobo.529.1">parents:</span></span></p>
<p class="calibre3"><span><span class="kobospan" id="kobo.530.1">[</span></span><span><span class="kobospan" id="kobo.531.1">p</span></span><span><span class="kobospan" id="kobo.532.1">a</span></span><span><span class="kobospan" id="kobo.533.1">r</span></span><span><span class="kobospan" id="kobo.534.1">e</span></span><span><span class="kobospan" id="kobo.535.1">n</span></span><span> </span><span><span class="kobospan" id="kobo.536.1">t</span></span><span><span class="kobospan" id="kobo.537.1"> </span></span><span><span class="kobospan" id="kobo.538.1">1</span></span><span> </span><span><span class="kobospan" id="kobo.539.1">−</span></span><span> </span><span><span class="kobospan" id="kobo.540.1">α</span></span><span><span class="kobospan" id="kobo.541.1">(</span></span><span><span class="kobospan" id="kobo.542.1">p</span></span><span><span class="kobospan" id="kobo.543.1">a</span></span><span><span class="kobospan" id="kobo.544.1">r</span></span><span><span class="kobospan" id="kobo.545.1">e</span></span><span><span class="kobospan" id="kobo.546.1">n</span></span><span> </span><span><span class="kobospan" id="kobo.547.1">t</span></span><span><span class="kobospan" id="kobo.548.1"> </span></span><span><span class="kobospan" id="kobo.549.1">2</span></span><span> </span><span><span class="kobospan" id="kobo.550.1">−</span></span><span> </span><span><span class="kobospan" id="kobo.551.1">p</span></span><span><span class="kobospan" id="kobo.552.1">a</span></span><span><span class="kobospan" id="kobo.553.1">r</span></span><span><span class="kobospan" id="kobo.554.1">e</span></span><span><span class="kobospan" id="kobo.555.1">n</span></span><span> </span><span><span class="kobospan" id="kobo.556.1">t</span></span><span><span class="kobospan" id="kobo.557.1"> </span></span><span><span class="kobospan" id="kobo.558.1">1</span></span><span><span class="kobospan" id="kobo.559.1">)</span></span><span><span class="kobospan" id="kobo.560.1">,</span></span><span> </span><span><span class="kobospan" id="kobo.561.1">p</span></span><span><span class="kobospan" id="kobo.562.1">a</span></span><span><span class="kobospan" id="kobo.563.1">r</span></span><span><span class="kobospan" id="kobo.564.1">e</span></span><span><span class="kobospan" id="kobo.565.1">n</span></span><span> </span><span><span class="kobospan" id="kobo.566.1">t</span></span><span><span class="kobospan" id="kobo.567.1"> </span></span><span><span class="kobospan" id="kobo.568.1">2</span></span><span> </span><span><span class="kobospan" id="kobo.569.1">+</span></span><span> </span><span><span class="kobospan" id="kobo.570.1">α</span></span><span><span class="kobospan" id="kobo.571.1">(</span></span><span><span class="kobospan" id="kobo.572.1">p</span></span><span><span class="kobospan" id="kobo.573.1">a</span></span><span><span class="kobospan" id="kobo.574.1">r</span></span><span><span class="kobospan" id="kobo.575.1">e</span></span><span><span class="kobospan" id="kobo.576.1">n</span></span><span> </span><span><span class="kobospan" id="kobo.577.1">t</span></span><span><span class="kobospan" id="kobo.578.1"> </span></span><span><span class="kobospan" id="kobo.579.1">2</span></span><span> </span><span><span class="kobospan" id="kobo.580.1">−</span></span><span> </span><span><span class="kobospan" id="kobo.581.1">p</span></span><span><span class="kobospan" id="kobo.582.1">a</span></span><span><span class="kobospan" id="kobo.583.1">r</span></span><span><span class="kobospan" id="kobo.584.1">e</span></span><span><span class="kobospan" id="kobo.585.1">n</span></span><span> </span><span><span><span class="kobospan" id="kobo.586.1">t</span></span></span><span><span><span class="kobospan" id="kobo.587.1"> </span></span></span><span><span><span class="kobospan" id="kobo.588.1">1</span></span></span><span><span><span class="kobospan" id="kobo.589.1">)</span></span></span><span><span><span class="kobospan" id="kobo.590.1">]</span></span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.591.1">The parameter, α, is a constant, whose value lies between 0 and 1. </span><span class="kobospan" id="kobo.591.2">With larger values of α, the interval </span><span><span class="kobospan" id="kobo.592.1">gets wider.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.593.1">For example, if the parents’ values are 1.33 and 5.72, the following will be </span><span><span class="kobospan" id="kobo.594.1">the case:</span></span></p>
<ul class="calibre10">
<li class="calibre11"><span class="kobospan" id="kobo.595.1">An α value of 0 will yield an interval of [1.33, 5.72] (similar to the interval between </span><span><span class="kobospan" id="kobo.596.1">the parents)</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.597.1">An α value of 0.5 will yield an interval of [-0.865, 7.915] (twice as wide as the interval between </span><span><span class="kobospan" id="kobo.598.1">the parents)</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.599.1">An α value of 1.0 will yield an interval of [-3.06, 10.11] (three times wider than the interval between </span><span><span class="kobospan" id="kobo.600.1">the parents)</span></span></li>
</ul>
<p class="calibre3"><span class="kobospan" id="kobo.601.1">These examples are illustrated in the following diagram, where the parents are labeled as </span><em class="italic"><span class="kobospan" id="kobo.602.1">p1</span></em><span class="kobospan" id="kobo.603.1"> and </span><em class="italic"><span class="kobospan" id="kobo.604.1">p2</span></em><span class="kobospan" id="kobo.605.1">, and the crossover interval is </span><span><span class="kobospan" id="kobo.606.1">colored yellow:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer034">
<span class="kobospan" id="kobo.607.1"><img alt="Figure 2.20: Blend crossover example" src="image/B20851_02_020.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.608.1">Figure 2.20: Blend crossover example</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.609.1">When using this crossover method, the α value is commonly set </span><span><span class="kobospan" id="kobo.610.1">to 0.5.</span></span></p>
<h2 id="_idParaDest-68" class="calibre7"><a id="_idTextAnchor082" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.611.1">Simulated binary crossover</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.612.1">The idea</span><a id="_idIndexMarker111" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.613.1"> behind </span><strong class="bold"><span class="kobospan" id="kobo.614.1">simulated binary crossover</span></strong><span class="kobospan" id="kobo.615.1"> (</span><strong class="bold"><span class="kobospan" id="kobo.616.1">SBX</span></strong><span class="kobospan" id="kobo.617.1">) is to imitate the</span><a id="_idIndexMarker112" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.618.1"> properties of the single-point crossover that is commonly used with binary-coded chromosomes. </span><span class="kobospan" id="kobo.618.2">One of these properties is that the average of the parents’ values is equal to that of the </span><span><span class="kobospan" id="kobo.619.1">offsprings’ values.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.620.1">When applying SBX, the two offspring are created from the two parents using the </span><span><span class="kobospan" id="kobo.621.1">following formula:</span></span></p>
<p class="calibre3"><span><span class="kobospan" id="kobo.622.1">o</span></span><span><span class="kobospan" id="kobo.623.1">f</span></span><span><span class="kobospan" id="kobo.624.1">f</span></span><span><span class="kobospan" id="kobo.625.1">s</span></span><span><span class="kobospan" id="kobo.626.1">p</span></span><span><span class="kobospan" id="kobo.627.1">r</span></span><span><span class="kobospan" id="kobo.628.1">i</span></span><span><span class="kobospan" id="kobo.629.1">n</span></span><span> </span><span><span class="kobospan" id="kobo.630.1">g</span></span><span><span class="kobospan" id="kobo.631.1"> </span></span><span><span class="kobospan" id="kobo.632.1">1</span></span><span> </span><span><span class="kobospan" id="kobo.633.1">=</span></span><span> </span><span><span class="kobospan" id="kobo.634.1"> </span></span><span><span class="kobospan" id="kobo.635.1">1</span></span><span><span class="kobospan" id="kobo.636.1"> </span></span><span><span class="kobospan" id="kobo.637.1">_</span></span><span><span class="kobospan" id="kobo.638.1"> </span></span><span><span class="kobospan" id="kobo.639.1">2</span></span><span><span class="kobospan" id="kobo.640.1"> </span></span><span><span class="kobospan" id="kobo.641.1">[</span></span><span><span class="kobospan" id="kobo.642.1">(</span></span><span><span class="kobospan" id="kobo.643.1">1</span></span><span> </span><span><span class="kobospan" id="kobo.644.1">+</span></span><span> </span><span><span class="kobospan" id="kobo.645.1">β</span></span><span><span class="kobospan" id="kobo.646.1">)</span></span><span><span class="kobospan" id="kobo.647.1">p</span></span><span><span class="kobospan" id="kobo.648.1">a</span></span><span><span class="kobospan" id="kobo.649.1">r</span></span><span><span class="kobospan" id="kobo.650.1">e</span></span><span><span class="kobospan" id="kobo.651.1">n</span></span><span> </span><span><span class="kobospan" id="kobo.652.1">t</span></span><span><span class="kobospan" id="kobo.653.1"> </span></span><span><span class="kobospan" id="kobo.654.1">1</span></span><span> </span><span><span class="kobospan" id="kobo.655.1">+</span></span><span> </span><span><span class="kobospan" id="kobo.656.1">(</span></span><span><span class="kobospan" id="kobo.657.1">1</span></span><span> </span><span><span class="kobospan" id="kobo.658.1">−</span></span><span> </span><span><span class="kobospan" id="kobo.659.1">β</span></span><span><span class="kobospan" id="kobo.660.1">)</span></span><span><span class="kobospan" id="kobo.661.1">p</span></span><span><span class="kobospan" id="kobo.662.1">a</span></span><span><span class="kobospan" id="kobo.663.1">r</span></span><span><span class="kobospan" id="kobo.664.1">e</span></span><span><span class="kobospan" id="kobo.665.1">n</span></span><span> </span><span><span class="kobospan" id="kobo.666.1">t</span></span><span><span class="kobospan" id="kobo.667.1"> </span></span><span><span class="kobospan" id="kobo.668.1">2</span></span><span><span class="kobospan" id="kobo.669.1">]</span></span></p>
<p class="calibre3"><span><span class="kobospan" id="kobo.670.1">o</span></span><span><span class="kobospan" id="kobo.671.1">f</span></span><span><span class="kobospan" id="kobo.672.1">f</span></span><span><span class="kobospan" id="kobo.673.1">s</span></span><span><span class="kobospan" id="kobo.674.1">p</span></span><span><span class="kobospan" id="kobo.675.1">r</span></span><span><span class="kobospan" id="kobo.676.1">i</span></span><span><span class="kobospan" id="kobo.677.1">n</span></span><span> </span><span><span class="kobospan" id="kobo.678.1">g</span></span><span><span class="kobospan" id="kobo.679.1"> </span></span><span><span class="kobospan" id="kobo.680.1">2</span></span><span> </span><span><span class="kobospan" id="kobo.681.1">=</span></span><span> </span><span><span class="kobospan" id="kobo.682.1"> </span></span><span><span class="kobospan" id="kobo.683.1">1</span></span><span><span class="kobospan" id="kobo.684.1"> </span></span><span><span class="kobospan" id="kobo.685.1">_</span></span><span><span class="kobospan" id="kobo.686.1"> </span></span><span><span class="kobospan" id="kobo.687.1">2</span></span><span><span class="kobospan" id="kobo.688.1"> </span></span><span><span class="kobospan" id="kobo.689.1">[</span></span><span><span class="kobospan" id="kobo.690.1">(</span></span><span><span class="kobospan" id="kobo.691.1">1</span></span><span> </span><span><span class="kobospan" id="kobo.692.1">−</span></span><span> </span><span><span class="kobospan" id="kobo.693.1">β</span></span><span><span class="kobospan" id="kobo.694.1">)</span></span><span><span class="kobospan" id="kobo.695.1">p</span></span><span><span class="kobospan" id="kobo.696.1">a</span></span><span><span class="kobospan" id="kobo.697.1">r</span></span><span><span class="kobospan" id="kobo.698.1">e</span></span><span><span class="kobospan" id="kobo.699.1">n</span></span><span> </span><span><span class="kobospan" id="kobo.700.1">t</span></span><span><span class="kobospan" id="kobo.701.1"> </span></span><span><span class="kobospan" id="kobo.702.1">1</span></span><span> </span><span><span class="kobospan" id="kobo.703.1">+</span></span><span> </span><span><span class="kobospan" id="kobo.704.1">(</span></span><span><span class="kobospan" id="kobo.705.1">1</span></span><span> </span><span><span class="kobospan" id="kobo.706.1">+</span></span><span> </span><span><span class="kobospan" id="kobo.707.1">β</span></span><span><span class="kobospan" id="kobo.708.1">)</span></span><span><span class="kobospan" id="kobo.709.1">p</span></span><span><span class="kobospan" id="kobo.710.1">a</span></span><span><span class="kobospan" id="kobo.711.1">r</span></span><span><span class="kobospan" id="kobo.712.1">e</span></span><span><span class="kobospan" id="kobo.713.1">n</span></span><span> </span><span><span class="kobospan" id="kobo.714.1">t</span></span><span><span class="kobospan" id="kobo.715.1"> </span></span><span><span class="kobospan" id="kobo.716.1">2</span></span><span><span class="kobospan" id="kobo.717.1">]</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.718.1">Here, β is a random number referred to as the spread factor. </span><span class="kobospan" id="kobo.718.2">This formula has the following </span><span><span class="kobospan" id="kobo.719.1">notable properties:</span></span></p>
<ul class="calibre10">
<li class="calibre11"><span class="kobospan" id="kobo.720.1">The average of the two offspring is equal to that of the parents, regardless of the value </span><span><span class="kobospan" id="kobo.721.1">of β</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.722.1">When the β value is 1, the offspring are duplicates of </span><span><span class="kobospan" id="kobo.723.1">the parents</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.724.1">When the β value is smaller than 1, the offspring are closer to each other than the </span><span><span class="kobospan" id="kobo.725.1">parents were</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.726.1">When the β value is larger than 1, the offspring are farther apart from each other than the </span><span><span class="kobospan" id="kobo.727.1">parents were</span></span></li>
</ul>
<p class="calibre3"><span class="kobospan" id="kobo.728.1">For example, if the</span><a id="_idIndexMarker113" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.729.1"> parents’ values are 1.33 and 5.72, the following</span><a id="_idIndexMarker114" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.730.1"> will be </span><span><span class="kobospan" id="kobo.731.1">the case:</span></span></p>
<ul class="calibre10">
<li class="calibre11"><span class="kobospan" id="kobo.732.1">A β value of 0.8 will yield 1.769 and 5.281 </span><span><span class="kobospan" id="kobo.733.1">as offspring</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.734.1">A β value of 1.0 will yield 1.33 and 5.72 </span><span><span class="kobospan" id="kobo.735.1">as offspring</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.736.1">A β value of 1.2 will yield 0.891 and 6.159 </span><span><span class="kobospan" id="kobo.737.1">as offspring</span></span></li>
</ul>
<p class="calibre3"><span class="kobospan" id="kobo.738.1">These cases are illustrated in the following diagram, where the parents are labeled as </span><em class="italic"><span class="kobospan" id="kobo.739.1">p1</span></em><span class="kobospan" id="kobo.740.1"> and </span><em class="italic"><span class="kobospan" id="kobo.741.1">p2</span></em><span class="kobospan" id="kobo.742.1"> and the offspring are labeled as </span><em class="italic"><span class="kobospan" id="kobo.743.1">o1</span></em> <span><span class="kobospan" id="kobo.744.1">and </span></span><span><em class="italic"><span class="kobospan" id="kobo.745.1">o2</span></em></span><span><span class="kobospan" id="kobo.746.1">:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer035">
<span class="kobospan" id="kobo.747.1"><img alt="      Figure 2.21: Simulated binary crossover example" src="image/B20851_02_021.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.748.1">      Figure 2.21: Simulated binary crossover example</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.749.1">In each of the</span><a id="_idIndexMarker115" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.750.1"> preceding cases, the average value of the two</span><a id="_idIndexMarker116" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.751.1"> offspring is 3.525, which is equal to the average value of the </span><span><span class="kobospan" id="kobo.752.1">two parents.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.753.1">Another property of binary single-point crossover that we would like to preserve is the similarity between offspring and parents. </span><span class="kobospan" id="kobo.753.2">This translates to the random distribution of the β value. </span><span class="kobospan" id="kobo.753.3">The probability of β should be much higher for values around 1, where the offspring are similar to the parents. </span><span class="kobospan" id="kobo.753.4">To achieve that, the β value is calculated using another random value, denoted by u, that is uniformly distributed over the interval [0, 1]. </span><span class="kobospan" id="kobo.753.5">Once the value of </span><em class="italic"><span class="kobospan" id="kobo.754.1">u</span></em><span class="kobospan" id="kobo.755.1"> is picked, β is calculated </span><span><span class="kobospan" id="kobo.756.1">as follows:</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.757.1">If </span><em class="italic"><span class="kobospan" id="kobo.758.1">u</span></em><span class="kobospan" id="kobo.759.1"> ≤ 0.5, we get the following: </span><span><span class="kobospan" id="kobo.760.1">β</span></span><span> </span><span><span class="kobospan" id="kobo.761.1">=</span></span><span> </span><span><span class="kobospan" id="kobo.762.1">(</span></span><span><span class="kobospan" id="kobo.763.1">2</span></span><span><span class="kobospan" id="kobo.764.1">u</span></span><span><span class="kobospan" id="kobo.765.1">)</span></span><span><span class="kobospan" id="kobo.766.1"> </span></span><span><span class="kobospan" id="kobo.767.1"> </span></span><span><span class="kobospan" id="kobo.768.1">1</span></span><span><span class="kobospan" id="kobo.769.1"> </span></span><span><span><span class="kobospan" id="kobo.770.1">_</span></span></span><span><span><span class="kobospan" id="kobo.771.1"> </span></span></span><span><span><span class="kobospan" id="kobo.772.1">1</span></span></span><span><span><span class="kobospan" id="kobo.773.1">−</span></span></span><span><span><span class="kobospan" id="kobo.774.1">η</span></span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.775.1">Otherwise, we get the following: </span><span><span class="kobospan" id="kobo.776.1">β</span></span><span> </span><span><span class="kobospan" id="kobo.777.1">=</span></span><span> </span><span><span class="kobospan" id="kobo.778.1">[</span></span><span><span class="kobospan" id="kobo.779.1"> </span></span><span><span class="kobospan" id="kobo.780.1">1</span></span><span><span class="kobospan" id="kobo.781.1"> </span></span><span><span class="kobospan" id="kobo.782.1">_</span></span><span><span class="kobospan" id="kobo.783.1"> </span></span><span><span class="kobospan" id="kobo.784.1">2</span></span><span><span class="kobospan" id="kobo.785.1">(</span></span><span><span class="kobospan" id="kobo.786.1">1</span></span><span> </span><span><span class="kobospan" id="kobo.787.1">−</span></span><span> </span><span><span class="kobospan" id="kobo.788.1">u</span></span><span><span class="kobospan" id="kobo.789.1">)</span></span><span><span class="kobospan" id="kobo.790.1">]</span></span><span><span class="kobospan" id="kobo.791.1"> </span></span><span><span class="kobospan" id="kobo.792.1"> </span></span><span><span class="kobospan" id="kobo.793.1">1</span></span><span><span class="kobospan" id="kobo.794.1"> </span></span><span><span><span class="kobospan" id="kobo.795.1">_</span></span></span><span><span><span class="kobospan" id="kobo.796.1"> </span></span></span><span><span><span class="kobospan" id="kobo.797.1">1</span></span></span><span><span><span class="kobospan" id="kobo.798.1">−</span></span></span><span><span><span class="kobospan" id="kobo.799.1">η</span></span></span></p>
<h2 id="_idParaDest-69" class="calibre7"><a id="_idTextAnchor083" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.800.1">Real mutation</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.801.1">One option for applying</span><a id="_idIndexMarker117" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.802.1"> mutation in real-coded genetic </span><a id="_idIndexMarker118" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.803.1">algorithms is to replace any real value with a brand-new one, generated randomly. </span><span class="kobospan" id="kobo.803.2">However, this can result in a mutated individual that has no relationship with the </span><span><span class="kobospan" id="kobo.804.1">original individual.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.805.1">Another approach is to generate a random real number that resides in the vicinity of the original individual. </span><span class="kobospan" id="kobo.805.2">An example of such a method is </span><a id="_idIndexMarker119" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.806.1">the </span><strong class="bold"><span class="kobospan" id="kobo.807.1">normally distributed</span></strong><span class="kobospan" id="kobo.808.1"> (or </span><strong class="bold"><span class="kobospan" id="kobo.809.1">Gaussian</span></strong><span class="kobospan" id="kobo.810.1">) </span><strong class="bold"><span class="kobospan" id="kobo.811.1">mutation</span></strong><span class="kobospan" id="kobo.812.1">: a random number is generated using a normal distribution with a mean value of zero and some predetermined standard deviation, as shown in the </span><span><span class="kobospan" id="kobo.813.1">following plot:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer036">
<span class="kobospan" id="kobo.814.1"><img alt="Figure 2.21: Gaussian mutation distribution example" src="image/B20851_02_022.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.815.1">Figure 2.21: Gaussian mutation distribution example</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.816.1">In the next two sections, we will go over a couple of advanced topics, namely </span><strong class="bold"><span class="kobospan" id="kobo.817.1">elitism</span></strong> <span><span class="kobospan" id="kobo.818.1">and </span></span><span><strong class="bold"><span class="kobospan" id="kobo.819.1">niching</span></strong></span><span><span class="kobospan" id="kobo.820.1">.</span></span></p>
<h1 id="_idParaDest-70" class="calibre5"><a id="_idTextAnchor084" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.821.1">Understanding elitism</span></h1>
<p class="calibre3"><span class="kobospan" id="kobo.822.1">While the average fitness of the </span><a id="_idIndexMarker120" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.823.1">genetic algorithm population generally increases as generations go by, it is possible that, at any point, the best individual(s) of the current generation will be lost. </span><span class="kobospan" id="kobo.823.2">This is due to the selection, crossover, and mutation operators altering the individuals in the process of creating the next generation. </span><span class="kobospan" id="kobo.823.3">In many cases, the loss is temporary as these individuals (or better individuals) will be re-introduced into the population in a </span><span><span class="kobospan" id="kobo.824.1">future generation.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.825.1">However, if we want to guarantee that the best individual(s) always make it to the next generation, we can apply the optional elitism strategy. </span><span class="kobospan" id="kobo.825.2">This means that the top </span><em class="italic"><span class="kobospan" id="kobo.826.1">n</span></em><span class="kobospan" id="kobo.827.1"> individuals (</span><em class="italic"><span class="kobospan" id="kobo.828.1">n</span></em><span class="kobospan" id="kobo.829.1"> being a small, predefined parameter) are duplicated into the next generation before we fill the rest of the available spots with offspring that are created using selection, crossover, and mutation. </span><span class="kobospan" id="kobo.829.2">The elite individuals who were duplicated are still eligible for the selection process, so they can still be used as the parents of </span><span><span class="kobospan" id="kobo.830.1">new individuals.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.831.1">Elitism </span><a id="_idIndexMarker121" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.832.1">can sometimes have a significant positive impact on the algorithm’s performance as it avoids the potential time waste needed for re-discovering good solutions that were lost in the </span><span><span class="kobospan" id="kobo.833.1">genetic flow.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.834.1">Another interesting way to enhance the results of genetic algorithms is the use of niching, as described in the </span><span><span class="kobospan" id="kobo.835.1">next section.</span></span></p>
<h1 id="_idParaDest-71" class="calibre5"><a id="_idTextAnchor085" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.836.1">Niching and sharing</span></h1>
<p class="calibre3"><span class="kobospan" id="kobo.837.1">In nature, all environments are </span><a id="_idIndexMarker122" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.838.1">further divided into multiple sub-environments, or niches, populated by various species taking advantage of the unique resources available in each niche, such </span><a id="_idIndexMarker123" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.839.1">as food and shelter. </span><span class="kobospan" id="kobo.839.2">For example, a forest environment is comprised of treetops, shrubs, the forest floor, tree roots, and so on; each of these accommodates different species that are specialized for living in that niche and take advantage of </span><span><span class="kobospan" id="kobo.840.1">its resources.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.841.1">When several different species coexist in the same niche, they all compete over the same resources, and a tendency is created to search for new, unpopulated niches and </span><span><span class="kobospan" id="kobo.842.1">populate them.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.843.1">In the realm of genetic algorithms, this niching phenomenon can be used to maintain the diversity of the population, as well as to find several optimal solutions, each considered </span><span><span class="kobospan" id="kobo.844.1">a niche.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.845.1">For example, suppose our genetic algorithm seeks to maximize a fitness function that has several peaks of varying heights, such as the one in the </span><span><span class="kobospan" id="kobo.846.1">following plot:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer037">
<span class="kobospan" id="kobo.847.1"><img alt="Figure 2.22: Expected genetic algorithm results without niching" src="image/B20851_02_023.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.848.1">Figure 2.22: Expected genetic algorithm results without niching</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.849.1">As the genetic algorithm</span><a id="_idIndexMarker124" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.850.1"> tends to find the global maximum, we expect, after a while, to see most of</span><a id="_idIndexMarker125" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.851.1"> the population concentrating around the top peak. </span><span class="kobospan" id="kobo.851.2">This is indicated in the preceding figure by the locations of the × marks on the function graph, representing individuals in the </span><span><span class="kobospan" id="kobo.852.1">current generation.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.853.1">However, there are implementations where, in addition to the global maximum, we would like to find some (or all) of the other peaks. </span><span class="kobospan" id="kobo.853.2">To make this happen, we could think of each peak as a niche, offering resources in the amount proportional to its height. </span><span class="kobospan" id="kobo.853.3">We then find a way to share (or divide) these resources among the individuals occupying them. </span><span class="kobospan" id="kobo.853.4">This will ideally drive the population to be distributed accordingly, with the top peak attracting the most individuals as it offers the most reward, and the other peaks populated with decreasing portions of the population as they offer smaller amounts </span><span><span class="kobospan" id="kobo.854.1">of reward.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.855.1">This ideal situation is depicted in the </span><span><span class="kobospan" id="kobo.856.1">following figure:</span></span></p>
<div class="calibre2">
<div class="img---figure" id="_idContainer038">
<span class="kobospan" id="kobo.857.1"><img alt="Figure 2.23: Ideal genetic algorithm results with niching" src="image/B20851_02_024.jpg" class="calibre4"/></span>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US"><span class="kobospan" id="kobo.858.1">Figure 2.23: Ideal genetic algorithm results with niching</span></p>
<p class="calibre3"><span class="kobospan" id="kobo.859.1">The challenge now is to </span><a id="_idIndexMarker126" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.860.1">implement this sharing mechanism. </span><span class="kobospan" id="kobo.860.2">One option to accomplish sharing is to </span><a id="_idIndexMarker127" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.861.1">divide the raw fitness value of each individual with (some function of) the combined distances from all the other individuals. </span><span class="kobospan" id="kobo.861.2">Another option would be to divide the raw fitness of each individual by the number of other individuals within a certain radius </span><span><span class="kobospan" id="kobo.862.1">around it.</span></span></p>
<h2 id="_idParaDest-72" class="calibre7"><a id="_idTextAnchor086" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.863.1">Serial niching versus parallel niching</span></h2>
<p class="calibre3"><span class="kobospan" id="kobo.864.1">Unfortunately, the </span><a id="_idIndexMarker128" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.865.1">niching concept, as described previously, can </span><a id="_idIndexMarker129" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.866.1">prove hard to implement as it increases the</span><a id="_idIndexMarker130" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.867.1"> complexity of the fitness calculation. </span><span class="kobospan" id="kobo.867.2">In practice, it </span><a id="_idIndexMarker131" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.868.1">will also require the population size to be the original one multiplied by the number of the expected peaks (which is </span><span><span class="kobospan" id="kobo.869.1">generally unknown).</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.870.1">One way to overcome these issues is to find the peaks one at a time (serial niching) instead of attempting to find all of them at the same time (parallel niching). </span><span class="kobospan" id="kobo.870.2">To implement serial niching, we use the genetic algorithm as usual and find the best solution. </span><span class="kobospan" id="kobo.870.3">Then, we update the fitness function so that the area of the maximum point that was found </span><a id="_idIndexMarker132" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.871.1">is flattened, and repeat the process of</span><a id="_idIndexMarker133" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.872.1"> the </span><a id="_idIndexMarker134" class="calibre6 pcalibre pcalibre1"/><span><span class="kobospan" id="kobo.873.1">genetic</span></span><span><a id="_idIndexMarker135" class="calibre6 pcalibre pcalibre1"/></span><span><span class="kobospan" id="kobo.874.1"> algorithm.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.875.1">Ideally, we will now find the next best peak as the original peak is no longer present. </span><span class="kobospan" id="kobo.875.2">We can repeat this process iteratively and find the next best peak at </span><span><span class="kobospan" id="kobo.876.1">each iteration.</span></span></p>
<h1 id="_idParaDest-73" class="calibre5"><a id="_idTextAnchor087" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.877.1">The art of solving problems using genetic algorithms</span></h1>
<p class="calibre3"><span class="kobospan" id="kobo.878.1">Genetic algorithms</span><a id="_idIndexMarker136" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.879.1"> provide us with a powerful and versatile tool that can be used to solve a wide array of problems and tasks. </span><span class="kobospan" id="kobo.879.2">When we set to work on a new problem, we need to customize the tool and match it to that problem. </span><span class="kobospan" id="kobo.879.3">This is done by making several choices, as described in the </span><span><span class="kobospan" id="kobo.880.1">following paragraphs.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.881.1">First, we need to determine </span><a id="_idIndexMarker137" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.882.1">the </span><strong class="bold"><span class="kobospan" id="kobo.883.1">fitness function</span></strong><span class="kobospan" id="kobo.884.1">. </span><span class="kobospan" id="kobo.884.2">This is how each individual will be evaluated, where larger values represent better individuals. </span><span class="kobospan" id="kobo.884.3">The function does not have to be mathematical. </span><span class="kobospan" id="kobo.884.4">It can be represented by an algorithm, a call to an external service, or even a result of a game played, to list a few options. </span><span class="kobospan" id="kobo.884.5">We just need a way to programmatically retrieve the fitness value for any given proposed </span><span><span class="kobospan" id="kobo.885.1">solution (individual).</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.886.1">Next, we need to choose an</span><a id="_idIndexMarker138" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.887.1"> appropriate </span><strong class="bold"><span class="kobospan" id="kobo.888.1">chromosome encoding</span></strong><span class="kobospan" id="kobo.889.1">. </span><span class="kobospan" id="kobo.889.2">This is based on the parameters we send to the fitness function. </span><span class="kobospan" id="kobo.889.3">So far, we have seen binary, integer, an ordered list, and real-coded examples. </span><span class="kobospan" id="kobo.889.4">However, for some problems, we may need a mix of parameter types, or may even decide to create our own custom </span><span><span class="kobospan" id="kobo.890.1">chromosome encoding.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.891.1">Next, we need to pick</span><a id="_idIndexMarker139" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.892.1"> a </span><strong class="bold"><span class="kobospan" id="kobo.893.1">selection</span></strong><span class="kobospan" id="kobo.894.1"> method. </span><span class="kobospan" id="kobo.894.2">Most selection methods will work for any kind of chromosome type. </span><span class="kobospan" id="kobo.894.3">If the fitness function is not directly accessible, but we still have a way to tell which of several candidate solutions is the best, we can consider utilizing the tournament </span><span><span class="kobospan" id="kobo.895.1">selection method.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.896.1">As we have seen in the preceding </span><a id="_idIndexMarker140" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.897.1">sections, the choice of </span><strong class="bold"><span class="kobospan" id="kobo.898.1">crossover</span></strong><span class="kobospan" id="kobo.899.1"> and </span><strong class="bold"><span class="kobospan" id="kobo.900.1">mutation</span></strong><span class="kobospan" id="kobo.901.1"> operators will be</span><a id="_idIndexMarker141" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.902.1"> linked to the chromosome encoding of the individuals. </span><span class="kobospan" id="kobo.902.2">Binary-coded chromosomes will have different crossover and mutation schemes than those that fit real-coded problems. </span><span class="kobospan" id="kobo.902.3">Similar to the choice of chromosome encoding, here, too, you can design methods for crossover and mutation that fit your unique </span><span><span class="kobospan" id="kobo.903.1">use case.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.904.1">Lastly, there are the hyperparameters</span><a id="_idIndexMarker142" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.905.1"> of the algorithm. </span><span class="kobospan" id="kobo.905.2">The most common parameter values we need to set are </span><span><span class="kobospan" id="kobo.906.1">as follows:</span></span></p>
<ul class="calibre10">
<li class="calibre11"><span><span class="kobospan" id="kobo.907.1">Population size</span></span></li>
<li class="calibre11"><span><span class="kobospan" id="kobo.908.1">Crossover rate</span></span></li>
<li class="calibre11"><span><span class="kobospan" id="kobo.909.1">Mutation rate</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.910.1">Max number </span><span><span class="kobospan" id="kobo.911.1">of generations</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.912.1">Other </span><span><span class="kobospan" id="kobo.913.1">stopping condition(s)</span></span></li>
<li class="calibre11"><span class="kobospan" id="kobo.914.1">Elitism (used or not; </span><span><span class="kobospan" id="kobo.915.1">what size)</span></span></li>
</ul>
<p class="calibre3"><span class="kobospan" id="kobo.916.1">For these parameters, we can choose what we deem as reasonable values and then tweak them, similar to how hyperparameters are dealt with in almost any other optimization and </span><span><span class="kobospan" id="kobo.917.1">learning algorithm.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.918.1">If making all these choices appears to be an overwhelming task, don’t fret! </span><span class="kobospan" id="kobo.918.2">In the chapters that follow, we will be iterating the process of making these choices time and again for the various types of problems we will tackle. </span><span class="kobospan" id="kobo.918.3">After reading this book, you will be able to look at new problems and make your own </span><span><span class="kobospan" id="kobo.919.1">wise choices.</span></span></p>
<h1 id="_idParaDest-74" class="calibre5"><a id="_idTextAnchor088" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.920.1">Summary</span></h1>
<p class="calibre3"><span class="kobospan" id="kobo.921.1">In this chapter, you were introduced to the basic flow of the genetic algorithm. </span><span class="kobospan" id="kobo.921.2">We then went over the key components of the flow, which included creating the population, calculating the fitness function, applying the genetic operators, and checking for </span><span><span class="kobospan" id="kobo.922.1">stopping conditions.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.923.1">Next, we went over various methods of selection, including roulette wheel selection, SUS, rank-based selection, fitness scaling, and tournament selection, and demonstrated the differences </span><span><span class="kobospan" id="kobo.924.1">between them.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.925.1">We continued by reviewing several crossover methods, including single-point, two-point, and k-point crossover, as well as OX1 and partially </span><span><span class="kobospan" id="kobo.926.1">matched crossover.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.927.1">You were then introduced to several mutation methods, including flip-bit mutation, followed by the swap, inversion, and </span><span><span class="kobospan" id="kobo.928.1">scramble mutations.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.929.1">Real-coded genetic algorithms were presented next, along with their specialized chromosome encoding and their custom genetic operators of crossover </span><span><span class="kobospan" id="kobo.930.1">and mutation.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.931.1">This was followed by an introduction to the concepts of elitism, niching, and sharing, as used in </span><span><span class="kobospan" id="kobo.932.1">genetic algorithms.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.933.1">In the last part of this chapter, you were presented with the various choices you need to make when approaching a problem to be solved using genetic algorithms, a procedure that will be repeated time and again throughout </span><span><span class="kobospan" id="kobo.934.1">this book.</span></span></p>
<p class="calibre3"><span class="kobospan" id="kobo.935.1">In the next chapter, the real fun begins – coding with Python! </span><span class="kobospan" id="kobo.935.2">You will be introduced to DEAP, an evolutionary computation framework that can be used as a powerful tool for applying genetic algorithms to a wide array of tasks. </span><span class="kobospan" id="kobo.935.3">DEAP will be used in the rest of this book as we develop Python programs that tackle numerous </span><span><span class="kobospan" id="kobo.936.1">different challenges.</span></span></p>
<h1 id="_idParaDest-75" class="calibre5"><a id="_idTextAnchor089" class="calibre6 pcalibre pcalibre1"/><span class="kobospan" id="kobo.937.1">Further reading</span></h1>
<p class="calibre3"><span class="kobospan" id="kobo.938.1">For more information, please refer to </span><a href="B20851_08.xhtml#_idTextAnchor238" class="calibre6 pcalibre pcalibre1"><span><em class="italic"><span class="kobospan" id="kobo.939.1">Chapter 8</span></em></span></a><span class="kobospan" id="kobo.940.1">, </span><em class="italic"><span class="kobospan" id="kobo.941.1">Genetic Algorithms</span></em><span class="kobospan" id="kobo.942.1">, in the book </span><em class="italic"><span class="kobospan" id="kobo.943.1">Artificial Intelligence with Python</span></em><span class="kobospan" id="kobo.944.1">, by Prateek Joshi, January 2017, available </span><span><span class="kobospan" id="kobo.945.1">at </span></span><a href="https://subscription.packtpub.com/book/big_data_and_business_intelligence/9781786464392/8" class="calibre6 pcalibre pcalibre1"><span><span class="kobospan" id="kobo.946.1">https://subscription.packtpub.com/book/big_data_and_business_intelligence/9781786464392/8</span></span></a><span><span class="kobospan" id="kobo.947.1">.</span></span></p>
</div>
</body></html>